server/map_rpc_server.go

```diff
@@ -321,6 +306,57 @@ func (t *TrillianMapServer) SetLeaves(ctx context.Context, req *trillian.SetMapL
 	return &trillian.SetMapLeavesResponse{MapRoot: newRoot}, nil
 }
 
+func doPreload(ctx context.Context, tx storage.MapTreeTX, hkv []merkle.HashKeyValue) error {
+	ctx, span := spanFor(ctx, "doPreload")
+	defer span.End()
+
+	readRev, err := tx.ReadRevision(ctx)
+	if err != nil {
+		return err
+	}
+	nidSet := make(map[string]bool)
+	type nodeAndID struct {
+		id   string
+		node storage.NodeID
+	}
+	c := make(chan nodeAndID, 2048)
+	var wg sync.WaitGroup
+
+	for _, i := range hkv {
+		wg.Add(1)
+		go func(k []byte) {
+			nid := storage.NewNodeIDFromHash(k)
+			sibs := (&nid).Siblings()
+			for _, sib := range sibs {
+				sibID := sib.String()
+				sib := sib
+				c <- nodeAndID{sibID, sib}
+			}
+			wg.Done()
+		}(i.HashedKey)
+	}
+
+	done := make(chan bool)
+	nids := make([]storage.NodeID, 0, len(hkv)*256)
```

Comment:
Consider extracting this 256 to a constant for SparseMerkleTreeDepth. If we ever wanted to have a Map of different depth then this would make it easier to evaluate where updates are needed. It also avoids the confusion I first had when reading this of assuming the 256 == 2^8 == number of leaves in a single stratum of the tree (I've spent some time today understanding NodeID, which may explain my incorrect first reading of this. YMMV).

Reply:
Passed in `hasher.BitLen()` for this for now, as this currently defines the height of the tree.